commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormEvaluator.java

```diff
@@ -0,0 +1,248 @@
+ * methods.
+ */
+public class EuclideanNormEvaluator {
+
+    /** Map of names to norm computation methods. */
+    private final Map<String, ToDoubleFunction<double[]>> methods = new LinkedHashMap<>();
+
+    /** Add a computation method to be evaluated.
+     * @param name method name
+     * @param method computation method
+     * @return this instance
+     */
+    public EuclideanNormEvaluator addMethod(final String name, final ToDoubleFunction<double[]> method) {
+        methods.put(name, method);
+        return this;
+    }
+
+    /** Evaluate the configured computation methods against the given array of input vectors.
+     * @param inputs array of input vectors
+     * @return map of evaluation results keyed by method name
+     */
+    public Map<String, Stats> evaluate(final double[][] inputs) {
+
+        final Map<String, StatsAccumulator> accumulators = new HashMap<>();
+        for (final String name : methods.keySet()) {
+            accumulators.put(name, new StatsAccumulator(inputs.length * 2));
+        }
+
+        for (int i = 0; i < inputs.length; ++i) {
+            // compute the norm in a forward and reverse directions to include
+            // summation artifacts
+            final double[] vec = inputs[i];
+
+            final double[] reverseVec = new double[vec.length];
+            for (int j = 0; j < vec.length; ++j) {
+                reverseVec[vec.length - 1 - j] = vec[j];
+            }
+
+            final double exact = computeExact(vec);
+
+            for (final Map.Entry<String, ToDoubleFunction<double[]>> entry : methods.entrySet()) {
+                final ToDoubleFunction<double[]> fn = entry.getValue();
+
+                final StatsAccumulator acc = accumulators.get(entry.getKey());
+
+                final double forwardSample = fn.applyAsDouble(vec);
+                acc.report(exact, forwardSample);
+
+                final double reverseSample = fn.applyAsDouble(reverseVec);
+                acc.report(exact, reverseSample);
+            }
+        }
+
+        final Map<String, Stats> stats = new LinkedHashMap<>();
+        for (final String name : methods.keySet()) {
+            stats.put(name, accumulators.get(name).computeStats());
+        }
+
+        return stats;
+    }
+
+    /** Compute the exact double value of the vector norm using BigDecimals
+     * with a math context of {@link MathContext#DECIMAL128}.
+     * @param vec input vector
+     * @return euclidean norm
+     */
+    private static double computeExact(final double[] vec) {
+        final MathContext ctx = MathContext.DECIMAL128;
+
+        BigDecimal sum = BigDecimal.ZERO;
+        for (final double v : vec) {
+            sum = sum.add(BigDecimal.valueOf(v).pow(2), ctx);
+        }
+
+        return sum.sqrt(ctx).doubleValue();
+    }
+
+    /** Compute the ulp difference between two values. This method assumes that
+     * the arguments have the same exponent.
```

Comment:
They do not have to have the same exponent. The ulp difference between 1.0 and Math.nextDown(1.0) will be 1. They do have to have the same sign. All your results should be positive so this is OK. 

If the numbers are very different then this will return a truncated difference by long to int primitive conversion. A return type of `long` would fix this. I do not think that you require `int` during testing as the sums use a double.

This comment also applies to `DoubleTestUtils` in the main package.